File: backend/agent/step_score.go
package agent

import (
	"encoding/json"
	"path/filepath"
	"regexp"
	"strings"

	"github.com/enough/enough/backend/agent/obligations"
)

var failurePathRE = regexp.MustCompile(`(?:(?:^|\s|\()([\w./-]+\.(?:go|py|rs|js|ts|tsx|jsx|java|c|h|cpp|hpp|rb|php|cs|swift|kt|scala|sql|yaml|yml|toml|json|md|sh))(?::\d+(?::\d+)?)?)|File "([^"]+)"`)

type stepTracker struct {
	lastBashCommand  string
	lastBashFailed   bool
	lastVerifyOutput string
	lastVerifyFailed bool
	failurePaths     []string
}

func (a *Agent) scoreToolStep(name, argsJSON string, result toolResult) *toolResult {
	if a.swarmDepth > 0 || !a.evidenceEnabled() || !a.cfg.Evidence.StepScorerEnabled() {
		a.recordStepOutcome(name, argsJSON, result)
		return nil
	}

	a.mu.Lock()
	tracker := a.step
	a.mu.Unlock()

	switch name {
	case "bash":
		cmd := bashCommandArg(argsJSON)
		if result.isErr && cmd != "" && tracker.lastBashFailed && cmd == tracker.lastBashCommand {
			return &toolResult{
				output: "REJECTED: repeated failing command — change approach or fix the failure site before re-running",
				isErr:  true,
			}
		}
	case "write_file", "edit_file":
		if !tracker.lastVerifyFailed || len(tracker.failurePaths) == 0 {
			break
		}
		path, ok := toolPathArg(argsJSON)
		if !ok {
			break
		}
		abs, err := a.resolvePath(path)
		if err != nil {
			break
		}
		if !editTouchesFailureSite(abs, tracker.failurePaths, a.evidenceLedger().MutatedPaths()) {
			return &toolResult{
				output: "REJECTED: edit does not touch the failure site — last verify failed on: " +
					strings.Join(tracker.failurePaths, ", "),
				isErr: true,
			}
		}
	}

	a.recordStepOutcome(name, argsJSON, result)
	return nil
}

func (a *Agent) recordStepOutcome(name, argsJSON string, result toolResult) {
	if name != "bash" {
		return
	}
	cmd := bashCommandArg(argsJSON)
	if cmd == "" {
		return
	}

	a.mu.Lock()
	defer a.mu.Unlock()

	a.step.lastBashCommand = cmd
	a.step.lastBashFailed = result.isErr

	reg := a.obligations
	isVerify := reg != nil && obligations.IsVerifyCommand(cmd, reg.VerifyCommand(), reg.ExtraVerifyCommands())
	if !isVerify {
		return
	}

	if result.isErr {
		a.step.lastVerifyFailed = true
		a.step.lastVerifyOutput = truncateStepOutput(result.output)
		a.step.failurePaths = extractFailurePaths(result.output, a.evidenceLedger().MutatedPaths())
	} else {
		a.step.lastVerifyFailed = false
		a.step.lastVerifyOutput = ""
		a.step.failurePaths = nil
	}
}

func bashCommandArg(argsJSON string) string {
	var args struct {
		Command string `json:"command"`
	}
	if err := json.Unmarshal([]byte(argsJSON), &args); err != nil {
		return ""
	}
	return strings.TrimSpace(args.Command)
}

func extractFailurePaths(output string, mutated []string) []string {
	seen := map[string]bool{}
	var paths []string
	add := func(p string) {
		p = strings.TrimSpace(p)
		if p == "" || seen[p] {
			return
		}
		seen[p] = true
		paths = append(paths, p)
	}
	for _, m := range failurePathRE.FindAllStringSubmatch(output, -1) {
		add(firstNonEmpty(m[1], m[2]))
	}
	for _, p := range mutated {
		add(filepath.ToSlash(p))
		add(filepath.Base(p))
	}
	return paths
}

func editTouchesFailureSite(editPath string, failurePaths, mutated []string) bool {
	editPath = filepath.Clean(editPath)
	for _, fp := range failurePaths {
		fp = filepath.Clean(fp)
		if fp == editPath || strings.HasSuffix(editPath, fp) || strings.HasSuffix(fp, filepath.Base(editPath)) {
			return true
		}
		if base := filepath.Base(fp); base != "" && strings.Contains(editPath, base) {
			return true
		}
	}
	for _, mp := range mutated {
		if filepath.Clean(mp) == editPath {
			return true
		}
	}
	return false
}

func truncateStepOutput(s string) string {
	const max = 4000
	if len(s) <= max {
		return s
	}
	return s[:max] + "\n... truncated ..."
}

File: backend/agent/step_score.ts
// PORT: backend/agent/step_score.go

import path from "node:path";
import { Effect } from "effect";
import { type toolResult } from "./agent";
import { Agent } from "./agent";
import { is_verify_command } from "./obligations/match";

const failurePathRE = /(?:(?:^|\s|\()([\w./-]+\.(?:go|py|rs|js|ts|tsx|jsx|java|c|h|cpp|hpp|rb|php|cs|swift|kt|scala|sql|yaml|yml|toml|json|md|sh))(?::\d+(?::\d+)?)?)|File "([^"]+)"/g;

export function scoreToolStep(
  this: Agent,
  name: string,
  argsJSON: string,
  result: toolResult
): toolResult | null {
  if (this.swarmDepth > 0 || !this.evidenceEnabled() || !this.cfg.evidence?.step_scorer) {
    this.recordStepOutcome(name, argsJSON, result);
    return null;
  }

  const tracker = this.step;

  switch (name) {
    case "bash": {
      const cmd = bashCommandArg(argsJSON);
      if (result.isErr && cmd !== "" && tracker.lastBashFailed && cmd === tracker.lastBashCommand) {
        return {
          output: "REJECTED: repeated failing command — change approach or fix the failure site before re-running",
          isErr: true,
        };
      }
      break;
    }
    case "write_file":
    case "edit_file": {
      if (!tracker.lastVerifyFailed || !tracker.failurePaths || tracker.failurePaths.length === 0) {
        break;
      }
      const [filePath, ok] = toolPathArg(argsJSON);
      if (!ok) {
        break;
      }
      // resolvePath returns an Effect. We run it synchronously using Effect.runSync.
      let abs = "";
      try {
        abs = Effect_runSync_resolvePath(this, filePath);
      } catch {
        break;
      }
      if (!editTouchesFailureSite(abs, tracker.failurePaths, this.evidenceLedger().mutated_paths())) {
        return {
          output: "REJECTED: edit does not touch the failure site — last verify failed on: " +
            tracker.failurePaths.join(", "),
          isErr: true,
        };
      }
      break;
    }
  }

  this.recordStepOutcome(name, argsJSON, result);
  return null;
}

Agent.prototype.scoreToolStep = scoreToolStep;

// Helper to run resolvePath synchronously
function Effect_runSync_resolvePath(agent: Agent, p: string): string {
  return Effect.runSync(agent.resolvePath(p));
}

export function recordStepOutcome(
  this: Agent,
  name: string,
  argsJSON: string,
  result: toolResult
): void {
  if (name !== "bash") {
    return;
  }
  const cmd = bashCommandArg(argsJSON);
  if (cmd === "") {
    return;
  }

  this.step.lastBashCommand = cmd;
  this.step.lastBashFailed = !!result.isErr;

  const reg = this.obligations;
  const isVerify = reg !== null && is_verify_command(cmd, reg.verify_command(), reg.extra_verify_commands());
  if (!isVerify) {
    return;
  }

  if (result.isErr) {
    this.step.lastVerifyFailed = true;
    this.step.lastVerifyOutput = truncateStepOutput(result.output);
    this.step.failurePaths = extractFailurePaths(result.output, this.evidenceLedger().mutated_paths());
  } else {
    this.step.lastVerifyFailed = false;
    this.step.lastVerifyOutput = "";
    this.step.failurePaths = null;
  }
}

Agent.prototype.recordStepOutcome = recordStepOutcome;

function bashCommandArg(argsJSON: string): string {
  let args: { command?: string };
  try {
    args = JSON.parse(argsJSON);
  } catch {
    return "";
  }
  return (args.command || "").trim();
}

function firstNonEmpty(a?: string, b?: string): string {
  if (a && a !== "") {
    return a;
  }
  return b || "";
}

function extractFailurePaths(output: string, mutated: string[]): string[] {
  const seen = new Set<string>();
  const paths: string[] = [];
  const add = (p: string) => {
    p = p.trim();
    if (p === "" || seen.has(p)) {
      return;
    }
    seen.add(p);
    paths.push(p);
  };

  failurePathRE.lastIndex = 0;
  const matches = [...output.matchAll(failurePathRE)];
  for (const m of matches) {
    add(firstNonEmpty(m[1], m[2]));
  }
  for (const p of mutated) {
    const toSlash = p.replaceAll("\\", "/");
    add(toSlash);
    add(path.basename(p));
  }
  return paths;
}

function editTouchesFailureSite(editPath: string, failurePaths: string[], mutated: string[]): boolean {
  const cleanEdit = path.normalize(editPath);
  for (const fp of failurePaths) {
    const cleanFp = path.normalize(fp);
    if (
      cleanFp === cleanEdit ||
      cleanEdit.endsWith(cleanFp) ||
      cleanFp.endsWith(path.basename(cleanEdit))
    ) {
      return true;
    }
    const base = path.basename(cleanFp);
    if (base !== "" && cleanEdit.includes(base)) {
      return true;
    }
  }
  for (const mp of mutated) {
    if (path.normalize(mp) === cleanEdit) {
      return true;
    }
  }
  return false;
}

function truncateStepOutput(s: string): string {
  const max = 4000;
  if (s.length <= max) {
    return s;
  }
  return s.slice(0, max) + "\n... truncated ...";
}

function toolPathArg(argsJSON: string): [string, boolean] {
  let args: { path?: string };
  try {
    args = JSON.parse(argsJSON);
  } catch {
    return ["", false];
  }
  if (!args.path || args.path === "") {
    return ["", false];
  }
  return [args.path, true];
}

/*
PORT STATUS
source path: backend/agent/step_score.go
source lines: 155
draft lines: 182
confidence: high
status: phase_b_compile
*/
